src/pool/srv_pool_check.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * (C) Copyright 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause-Patent
+ */
+/**
+ * \file
+ *
+ * ds_pool: Pool Service Check
+ */
+
+#define D_LOGFAC DD_FAC(pool)
+
+#include <daos_srv/pool.h>
+
+#include <sys/stat.h>
+#include <daos_srv/daos_engine.h>
+#include <daos_srv/daos_mgmt_srv.h>
+#include <daos_srv/rdb.h>
+#include "srv_internal.h"
+#include "srv_layout.h"
+
+static int
+pool_svc_glance(uuid_t uuid, char *path, struct ds_pool_svc_clue *clue_out)
+{
+	struct rdb_storage     *storage;
+	struct ds_pool_svc_clue	clue;
+	struct rdb_tx		tx;
+	rdb_path_t		root;
+	struct pool_buf	       *map_buf;
+	int			rc;
+
+	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
+
+	rc = rdb_open(path, uuid, NULL /* cbs */, NULL /* arg */, &storage);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to open %s: "DF_RC"\n", DP_UUID(uuid), path, DP_RC(rc));
+		goto out;
+	}
+
+	rc = rdb_glance(storage, &clue.psc_db_clue);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to glance at %s: "DF_RC"\n", DP_UUID(uuid), path,
+			DP_RC(rc));
+		goto out_storage;
+	}
+
+	rc = rdb_tx_begin_local(storage, &tx);
+	if (rc != 0)
+		goto out_db_clue;
+
+	rc = rdb_path_init(&root);
+	if (rc != 0)
+		goto out_tx;
+	rc = rdb_path_push(&root, &rdb_path_root_key);
+	if (rc != 0)
+		goto out_root;
+
+	rc = ds_pool_svc_load(&tx, uuid, &root, &map_buf, &clue.psc_map_version);
+	if (rc == DER_UNINIT) {
+		clue.psc_map_version = 0;
+		rc = 0;
+	} else if (rc != 0) {
+		goto out_root;
+	}
+
+	*clue_out = clue;
+	D_FREE(map_buf);
+out_root:
+	rdb_path_fini(&root);
+out_tx:
+	rdb_tx_end(&tx);
+out_db_clue:
+	if (rc != 0)
+		d_rank_list_free(clue.psc_db_clue.bcl_replicas);
+out_storage:
+	rdb_close(storage);
+out:
+	return rc;
+}
+
+/**
+ * Glance at the pool with \a uuid in \a dir, and report \a clue about its
+ * persistent state. Note that if an error has occurred, it is reported in \a
+ * clue->pc_rc, with \a clue->pc_uuid and \a clue->pc_dir fields also being
+ * valid.
+ *
+ * \param[in]	uuid	pool UUID
+ * \param[in]	dir	storage directory
+ * \param[out]	clue	pool clue
+ */
+void
+ds_pool_glance(uuid_t uuid, enum ds_pool_dir dir, struct ds_pool_clue *clue)
+{
+	char	       *path;
+	struct stat	st;
+	int		rc;
+
+	memset(clue, 0, sizeof(*clue));
+	uuid_copy(clue->pc_uuid, uuid);
+	clue->pc_rank = dss_self_rank();
```

Comment:
in the rdb_clue related code (rdb_glance() function) there is a statement about how, at least at the rdb level, the rank should be stored persistently in the future. What is the situation with ordinary, non PS replica pool storage engines (I forget about whether and how rank reuse is achieved). Is a similar comment useful here?

Reply:
We store the rank of an engine, if one has been assigned by the MS and received by the local server, persistently in <storage_path>/superblock. The server reads the rank from the superblock file and sets the value to the engine in the engine startup process. Hence, I think we don't need a similar comment here.

When writing the comment in rdb_clue, I think I was considering future RDB node IDs that could differ from ranks (<rank, generate> or a completely unrelated number). I'll change the comment.